### PR TITLE
rewrite a lot of hellabot in a cleaner, more idiomatic style

### PR DESCRIPTION
**Pull request by whyrusleeping**
This is a squash and rebase of @flexd's PR. 

@flexd want to give this a look over?

### CODE REVIEW

**Comment by flexd** [comment]
@whyrusleeping Looking at it, but it's a bit messy to see because of all the different files changed a few times.

So far I see we need to fix a few things, mostly related to golint, http://goreportcard.com/report/flexd/hellabot (see here), and also the examples I think.

**Comment by whyrusleeping** [comment]
@flexd I can do the golint. I'll also give you collab here.

File: hellabot.go [hunk 1]
```diff
@@ -2,232 +2,175 @@ package hbot
...
-// load a Json config file
-func LoadConfig(f string) (Config, error) {
-
-	file, err := os.Open(f)
-
-	if err != nil {
-		fmt.Println("Couldn't read config file")
-		return Config{}, err
+// NewBot creates a new instance of Bot
+func NewBot(host, nick string, options ...func(*Bot)) (*Bot, error) {
```
Comment (whyrusleeping) [inline hellabot.go]: @flexd didnt notice this one. what is the proper use of the options now?

Comment (flexd) [inline hellabot.go]: Look at http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

They are used like this

```go
hijackSession := func(bot *hbot.Bot) {                                                                                                                                                                                                                                                 
        bot.HijackSession = true                                                                                                                                                                                                                                                           
    }                                                                                                                                                                                                                                                                                      
    channels := func(bot *hbot.Bot) {                                                                                                                                                                                                                                                      
        bot.Channels = []string{"#test"}                                                                                                                                                                                                                                                   
    }                                                                                                                                                                                                                                                                                      
    irc, err := hbot.NewBot(*serv, *nick, hijackSession, channels)                                                                                                                                                                                                                         
    if err != nil {                                                                                                                                                                                                                                                                        
        panic(err)                                                                                                                                                                                                                                                                         
    }                                                               
```

You could use as many options as you wish, or you could do many changes in one option. I plan to follow the example in the link a bit closer and return an error at some point, but currently we have no options that can fail so there is no point.

**Comment by flexd** [comment]
I think this looks good to me, for now. Want to merge it? We should probably tag this as a release, and update the examples.

I want to discuss changing (or adding) how triggers work to maybe use a Interface ala. http.Handler as we have talked about in other issues, but that can come later. There would be several things we would have to solve in order to do that I think.

**Comment by whyrusleeping** [comment]
I think the triggers should be kept as is for now, that change can be discussed in a later PR, i'm fine merging this without those for now. Also, the functional options thing is quite nice!

**Comment by flexd** [comment]
This has worked for me for the past month or so without issues. I'm merging it :+1:

**Comment by whyrusleeping** [comment]
@flexd sounds good! i was just writing a new bot last night, nice timing :)

**Pull request merged**